src/watertap_contrib/reflo/unit_models/zero_order/tests/test_crystallizer_watertap.py

```diff
@@ -0,0 +1,787 @@
+            "eq_mass_frac_phase_comp",
+            "eq_dens_mass_solvent",
+            "eq_dens_mass_solute",
+            "eq_dens_mass_phase",
+            "eq_cp_mass_solute",
+            "eq_cp_mass_phase",
+            "eq_flow_vol_phase",
+            "eq_enth_mass_solvent",
+        ]
+        # List of attributes for liquid stateblocks only
+        stateblock_objs_liq_lst = ["pressure_sat", "eq_pressure_sat"]
+        # Inlet block
+        assert hasattr(m.fs.unit, "properties_in")
+        blk = getattr(m.fs.unit, "properties_in")
+        for var_str in stateblock_objs_lst:
+            assert hasattr(blk[0], var_str)
+        for var_str in stateblock_objs_liq_lst:
+            assert hasattr(blk[0], var_str)
+
+        # Liquid outlet block
+        assert hasattr(m.fs.unit, "properties_out")
+        blk = getattr(m.fs.unit, "properties_out")
+        for var_str in stateblock_objs_lst:
+            assert hasattr(blk[0], var_str)
+        for var_str in stateblock_objs_liq_lst:
+            assert hasattr(blk[0], var_str)
+
+        # Vapor outlet block
+        assert hasattr(m.fs.unit, "properties_vapor")
+        blk = getattr(m.fs.unit, "properties_vapor")
+        for var_str in stateblock_objs_lst:
+            assert hasattr(blk[0], var_str)
+
+        # Liquid outlet block
+        assert hasattr(m.fs.unit, "properties_solids")
+        blk = getattr(m.fs.unit, "properties_solids")
+        for var_str in stateblock_objs_lst:
+            assert hasattr(blk[0], var_str)
+
+        # test statistics
+        assert number_variables(m) == 255
+        assert number_total_constraints(m) == 138
+        assert number_unused_variables(m) == 5
+
+    @pytest.mark.unit
+    def test_dof(self, Crystallizer_frame):
+        m = Crystallizer_frame
+        assert degrees_of_freedom(m) == 0
+
+    @pytest.mark.unit
+    def test_calculate_scaling(self, Crystallizer_frame):
+        m = Crystallizer_frame
+
+        m.fs.properties.set_default_scaling(
+            "flow_mass_phase_comp", 1e-1, index=("Liq", "H2O")
+        )
+        m.fs.properties.set_default_scaling(
+            "flow_mass_phase_comp", 1e-1, index=("Liq", "NaCl")
+        )
+        m.fs.properties.set_default_scaling(
+            "flow_mass_phase_comp", 1e-1, index=("Vap", "H2O")
+        )
+        m.fs.properties.set_default_scaling(
+            "flow_mass_phase_comp", 1e-1, index=("Sol", "NaCl")
+        )
+        calculate_scaling_factors(m)
+
+        # check that all variables have scaling factors
+        unscaled_var_list = list(unscaled_variables_generator(m))
+        assert len(unscaled_var_list) == 0
+
+        for _ in badly_scaled_var_generator(m):
+            assert False
+
+    @pytest.mark.requires_idaes_solver
+    @pytest.mark.component
+    def test_initialize(self, Crystallizer_frame):
+        # Add costing function, then initialize
+        m = Crystallizer_frame
+        # m.fs.costing = WaterTAPCosting()
```

Comment:
@Zhuoran29 just skimming this from my phone, but maybe the failed tests are due to the costing lines being commented out here.

Reply:
In any case, I’d recommend just having a separate test_costing test rather than bundle it up in the initialization test.

Reply:
Now the costing package is rebuilt in REFLO and tested separately.